src/compile/layer.ts

```diff
@@ -89,39 +87,62 @@ export class LayerModel extends Model {
       // TODO: correctly implement independent scale
       if (true) { // if shared/union scale
         keys(child.component.scale).forEach(function(channel) {
-          let newScales = child.component.scale[channel] as ScaleComponent[];
-          if (!newScales) {
+          let childScales: ScaleComponents = child.component.scale[channel];
+          if (!childScales) {
+            // the child does not have any scales so we have nothing to merge
             return;
           }
 
-          const existingScales = scaleComponent[channel];
-          if (existingScales && newScales[0].type === ScaleType.ORDINAL) {
-            // Ordinal scales are unioned by combining the domain of the first scale. Other scales are appended.
-            const existingDomain = existingScales[0].domain.fields ? existingScales[0].domain : { fields: [existingScales[0].domain] };
-            newScales[0].domain = {
-              // assumes that the new scale domain is not a list
-              fields: existingDomain.fields.concat([newScales[0].domain])
-            };
-            if (existingScales.length > 1) {
-              delete existingScales[0];
-              scaleComponent[channel] = newScales.concat(existingScales);
+          const modelScales: ScaleComponents = scaleComponent[channel];
+          if (modelScales && modelScales.main) {
+            // Scales are unioned by combining the domain of the first scale (the main scale).
+            // Other scales that are used for ordinal legends are appended.
+            const modelDomain = modelScales.main.domain;
+            const childDomain = childScales.main.domain;
+
+            if (isArray(modelDomain)) {
+              if (isArray(childScales.main.domain)) {
+                modelScales.main.domain = modelDomain.concat(childDomain);
+              } else {
+                model.addWarning('scales cannot be unioned with explicit domain');
```

Comment:
Do you mean

"custom domain scale cannot be unioned with default field-based domain?"